### DISCUSSION TITLE
cuQuantum Performance Benchmark Suite v0.2.0 is now live!

### DISCUSSION THREAD

**Discussion by tlubowe** [NVIDIA/cuQuantum#41] [Announcements]
Hello cuQuantum users, we are happy to share the release of NVIDIA cuQuantum Performance Benchmark Suite v0.2.0 🚀 

## Breaking Changes

- To run the circuit-level benchmarks, a `circuit` subcommand needs to be added. For example,
```bash
cuquantum-benchmarks --frontend cirq --backend qsim --benchmark qft --nqubits 16  # v0.1.0
```
becomes
```bash
cuquantum-benchmarks circuit --frontend cirq --backend qsim --benchmark qft --nqubits 16  # v0.2.0
```

## New Features & Enhancements

- Introduce API-level benchmarks. Currently, `cuquantum.custatevec.apply_matrix()` is supported. See `cuquantum-benchmarks api --help` for more details.
   - The existing circuit-level benchmarks should be run with the addition of the subcommand `circuit`, as shown above.
- Support for more simulator frontends & backends, including Pennylane and Qulacs.
- Support single- and multi-process Slurm jobs.
- Add more command line options to support the cusvaer backend.
- Include more system & runtime information in the generated json logs.
- Improve the internal timer implementation.
- Improve the overall code quality, such as the logging system and the circuit generator.
   - The cached circuit data in `/path/to/your/cachedir/circuit/` would be automatically invalidated. 

## Bugs fixed

- Work around a bug in cusvaer 22.11.
- Fix potential circuit mismatches across multiple processes.
- Fix potential random seed mismatches across multiple processes.
- Fix the unexpected behavior of `--cachedir`.

## Installation Guide

Same as in the previous release, it can be easily installed by cloning this repository as follows:
```bash
$ git clone https://github.com/NVIDIA/cuQuantum.git
$ cd cuQuantum/benchmarks/
$ pip install .
$ cuquantum-benchmarks --help
```
This benchmark suite works either without or with our [cuQuantum Appliance container](https://docs.nvidia.com/cuda/cuquantum/appliance/index.html). For further information please see the [benchmark project README](https://github.com/NVIDIA/cuQuantum/blob/main/benchmarks/README.md). 

## Acknowledgement

Special shoutout for @Hosseinberg for contributing to this project 🙏

Happy benchmarking!

**Comment by rht** [comment]
Thank you for improving the benchmark.
I found some `cutn` results to be slower than they should be due to `samples` being hardcoded to be 512 https://github.com/NVIDIA/cuQuantum/blob/2cc68ede865d1f2028f28c9ad411f85748e20b58/benchmarks/cuquantum_benchmarks/backends/backend_cutn.py#L107. Would it be better to not specify it at all?

**Reply by leofang** [reply]
Thanks for raising this concern @rht. We (w/ @haidarazzam) already had some ideas for relaxing this hard coded value (internal issue 34). Since you've already known how to find this line and you need to build `cuquantum-benchmarks` from source anyway, I suggest to work around it by removing `samples` here for the time being. We'll improve it in a future release 🙂

**Reply by leofang** [reply]
FYI in cuquantum-benchmarks v0.3.0 (just pushed) we now offer a CLI option to adjust `samples`, though if you use cuQuantum 23.06+ there should also be some speedup already.
